ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandStore.java

```diff
@@ -186,6 +214,37 @@ public QueuedCommandStatus enqueueCommand(
     return restoreCommands;
   }
 
+  @Override
+  public void ensureConsumedUpThrough(final long seqNum, final long timeout)
+      throws TimeoutException {
+    final long consumerPosition = getNextConsumerSequenceNumber();
+    if (consumerPosition > seqNum) {
```

Comment:
I think this is not quite what we want here. `consumerPosition` is the consumer's position, which is the offset of the next record to be consumed. What we want here is the offset of the last command that we know we processed. These are not the same for the current batch of records being processed. Imagine the executor is off processing a batch of records consisting of a single record w/ offset 10. consumerPosition would be 11, so we might not wait for `seqNum=10`, when we should, since we don't know if its been completed yet. So what we really want is our own bookkeeping of the latest processed command.

I also think you should move this bookkeeping and check into SequenceNumberFutureStore and add some synchronization with `completeFuturesUpToSequenceNumber`. The reason is that there's a small race here where an API thread could call `ensureConsumedUpThrough` and the statement executor thread could be just finishing the latest batch of statements. If it completes futures and sleeps on consume between the check and registering the future, this could timeout when it shouldn't.

Reply:
Good catch! Done.